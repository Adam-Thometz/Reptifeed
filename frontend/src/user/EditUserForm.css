.EditUserForm {
  border: 2px solid #D5CE3B;
  width: 50vw;
  height: 60%;
  border-radius: 20px;
  margin: 0 auto;
  padding: 1%;
  background-color: #3E3E07;
  color: #D5CE3B;
  box-shadow: 10px 5px 5px rgb(4, 66, 4);
}

.EditUserForm-form-group {
  width: 70%;
  margin: 0 auto;
  padding: 10px;
}

.EditUserForm-label {
  width: 10%;
  margin: 0 1%;
}

.EditUserForm-input {
  width: 75%;
  padding: 1%;
  border-radius: 5px;
  border: none;
}

.EditUserForm-submit {
  padding: 10px;
  border-radius: 10px;
  border: 1px solid #D5CE3B;
  background-color: #AA7B08;
  font-size: 1rem;
}