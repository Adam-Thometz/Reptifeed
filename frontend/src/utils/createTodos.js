
/** createTodos(reptiles, pantry)
 * 
 * Takes a user's reptiles and pantry and creates a list of things they need to do to unlock the app's full features
 * 
 * Returns an two arrays of Link components, one of important things and one of nice-to-haves
 */

function createTodos(reptiles, pantry) {
  /** Essential checks:
   * - reptileCheck: Checks if a user has reptiles
   * - vegetableOftenCheck: Checks if a user has a vegetable staple in their pantry
   * - proteinOftenCheck: Checks if a user has a protein staple in their pantry
   * - treatModeratelyCheck: Checks if a user has a treat staple in their pantry
   * - supplementCheck: Checks if a user has at least two different supplements
   */
  const essentialTodos = [];

  const reptileCheck = !!reptiles.length;
  const vegetableOftenCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'often');
  const proteinOftenCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'often');
  const treatModeratelyCheck = pantry.some(food => food.isTreat && food.frequency === 'moderately');
  const supplementCheck = (pantry.filter(food => food.type === 'supplement')).length >= 2;

  if (!reptileCheck) essentialTodos.push('Add a reptile');
  if (!vegetableOftenCheck) essentialTodos.push('Add vegetables eaten often');
  if (!proteinOftenCheck) essentialTodos.push('Add protein eaten often');
  if (!treatModeratelyCheck) essentialTodos.push('Add a treat eaten moderately');
  if (!supplementCheck) essentialTodos.push('Add at least two supplements');

  /** Nice-to-have checks:
   * - vegetableModeratelyCheck: Checks if a user has a vegetable that's eaten moderately in their pantry
   * - vegetableOccasionallyCheck: Checks if a user has a vegetable that's eaten occasionally in their pantry
   * - proteinModeratelyCheck: Checks if a user has a protein that's eaten moderately in their pantry
   * - proteinOccasionallyCheck: Checks if a user has a protein that's eaten occasionally in their pantry
   * - treatOccasionallyCheck: Checks if a user has a treat that's eaten rarely in their pantry
   */
  const niceToHaveTodos = [];

  const vegetableModeratelyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'moderately');
  const vegetableOccasionallyCheck = pantry.some(food => food.type === 'vegetable' && food.frequency === 'occasionally' && !food.isTreat);
  const proteinModeratelyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'moderately');
  const proteinOccasionallyCheck = pantry.some(food => food.type === 'protein' && food.frequency === 'occasionally' && !food.isTreat);
  const treatOccasionallyCheck = pantry.some(food => food.isTreat && food.frequency === 'occasionally');

  if (!vegetableModeratelyCheck) niceToHaveTodos.push('Add a vegetable eaten moderately');
  if (!vegetableOccasionallyCheck) niceToHaveTodos.push('Add a vegetable eaten occasionally');
  if (!proteinModeratelyCheck) niceToHaveTodos.push('Add a protein eaten moderately');
  if (!proteinOccasionallyCheck) niceToHaveTodos.push('Add a protein eaten occasionally');
  if (!treatOccasionallyCheck) niceToHaveTodos.push('Add a treat eaten occasionally');

  return {essentialTodos, niceToHaveTodos};
};

export default createTodos;