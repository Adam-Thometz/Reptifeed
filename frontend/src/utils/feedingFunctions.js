/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
*/

function getFoodFromPantry(pantry, freq, type) {
  const availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) return `Get some more ${type}`
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  const availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) return "You don't have any treats";
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

export { getFoodFromPantry, getTreatFromPantry };