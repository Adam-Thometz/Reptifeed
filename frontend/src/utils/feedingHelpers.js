/** Feeding functions
 * 
 * Used to get foods from pantry that match frequencies and food type.
 * 
 * If no foods are available for the frequency, returns foods eaten most often
*/

function getFoodFromPantry(pantry, freq, type) {
  let availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === freq && food.type === type)));
  if (!availableFood.length) {
    availableFood = pantry.filter(food => (!food.isTreat && (food.frequency === 'often' && food.type === type)));
  };
  const selectedIdx = Math.floor(Math.random() * availableFood.length);

  return availableFood[selectedIdx];
};

function getTreatFromPantry(pantry, freq) {
  let availableTreats = pantry.filter(food => food.isTreat && food.frequency === freq);
  if (!availableTreats.length) {
    availableTreats = pantry.filter(food => food.isTreat);
  };
  const selectedIdx = Math.floor(Math.random() * availableTreats.length); 
  
  return availableTreats[selectedIdx];
};

function getNextSupplement(pantry, lastSupplement) {
  const nextSupplement = lastSupplement === 'calcium' ?
    pantry.find(food => food.name === 'multivitamin') :
    pantry.find(food => food.name === 'calcium');
  
  return nextSupplement;
}

export { getFoodFromPantry, getTreatFromPantry, getNextSupplement };