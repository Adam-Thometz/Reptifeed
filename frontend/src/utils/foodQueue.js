/** Queue: a linked-list data structure used to represent feeding patterns.
 * 
 * A simple queue used to cycle through frequencies to determine what to feed your reptile.
 * 
 * Takes the frequency at the head and returns it to the back once we're done.
 * 
 * Also used to determine what treats to serve.
*/

// Node class to be used for the queue
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
  };
};

// Queue class. A line of frequencies to determine what to feed reptile
class Queue {
  constructor(vals) {
    this.first = null;
    this.last = null;

    for (let val of vals) this.enqueue(val)
  };

  // enqueue(val): add to end of queue. Returns undefined.
  enqueue(val) {
    const node = new Node(val);
    if (!this.first) {
      this.first = node;
      this.last = node;
    } else {
      this.last.next = node;
      this.last = node;
    };
  };

  // dequeue(): remove from beginning of queue. Returns val.
  dequeue() {
    const val = this.first.val;
    this.first = this.first.next;
    return val;
  };
};

// foodQueue: Feed reptile based on frequency
const foodQueue = new Queue(["often", "often", "moderately", "often", "often", "moderately", "often", "often", "moderately", "occasionally"]);

// treatQueue: Give reptile treats based on frequency
const treatQueue = new Queue(["moderately", "moderately", "occasionally"]);

export { foodQueue, treatQueue };