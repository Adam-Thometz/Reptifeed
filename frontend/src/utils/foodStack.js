// foodStack: Feed reptile based on frequency
const foodStack = ["occasionally", "moderately", "often", "often", "moderately", "often", "often", "moderately", "often", "often"];

// treatStack: Give reptile treats based on frequency
const treatStack = ["occasionally", "moderately", "moderately"];

export { foodStack, treatStack };