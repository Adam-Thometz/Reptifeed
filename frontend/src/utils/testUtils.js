import React from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: 'rickSanchez',
  password: 'ilikeMorty',
  email: 'rickman@qmail.com',
  isAdmin: false
};

const demoReptile = [
  {
    id: 1,
    name: 'two crows',
    species: 'crows',
    subspecies: 'two of them',
    birthday: '2015-10-29',
    imgUrl: '',
    ownerId: 1
  }
];

const demoPantry = [
  {
    name: 'plumbus',
    type: 'everything',
    frequency: 'always',
    image: '',
    isTreat: true,
    tips: "Everyone needs a plumbus!"
  }
];

const UserProvider = ({ children, currUser = demoUser, pantry = demoPantry, reptiles = demoReptile }) => (
  <UserContext.Provider value={{currUser, pantry, reptiles}}>
    {children}
  </UserContext.Provider>
);

export { UserProvider };