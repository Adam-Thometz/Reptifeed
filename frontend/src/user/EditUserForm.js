import React, { useContext, useState } from "react";
import { useNavigate, useParams } from 'react-router-dom';
import ReptifeedApi from "../api";
import Alert from "../common/Alert";
import UserContext from "../utils/UserContext";
import './EditUserForm.css'

const EditUserForm = () => {
  const { currUser, setCurrUser } = useContext(UserContext)
  const { id } = useParams();
  const navigate = useNavigate();

  document.title = `Edit ${currUser.username}'s Profile | Reptifeed`;
  const init = {
    username: currUser.username,
    password: '',
    email: currUser.email
  }; 

  const [formData, setFormData] = useState(init);
  const [formMessages, setFormMessages] = useState([]);

  const handleChange = e => {
    const { name, value } = e.target;
    setFormData(d => ({ ...d, [name]: value }));
  };

  const handleSubmit = async e => {
    e.preventDefault();
    try {
      const res = await ReptifeedApi.updateUser(+id, formData);
      setCurrUser(res);
      navigate(`/users/${res.id}`);
    } catch (errors) {
      setFormMessages(m => [...m, errors]);
    };
  };

  return (
    <div className="EditUserForm">
      <h1>Edit your information</h1>
      <form onSubmit={handleSubmit}>
        <div className="EditUserForm-form-group">
          <label className="EditUserForm-label" htmlFor="username">Username:</label>
          <input
            className="EditUserForm-input"
            type='text'
            id='username'
            name='username'
            value={formData.username}
            onChange={handleChange}
          />
        </div>
        <div className="EditUserForm-form-group">
          <label className="EditUserForm-label" htmlFor="password">Password:</label>
          <input
            className="EditUserForm-input"
            type='password'
            id='password'
            name='password'
            value={formData.password}
            onChange={handleChange}
          />
        </div>
        <div className="EditUserForm-form-group">
          <label className="EditUserForm-label" htmlFor="email">Email:</label>
          <input
            className="EditUserForm-input"
            type='email'
            id='email'
            name='email'
            value={formData.email}
            onChange={handleChange}
          />
        </div>
        <button className="EditUserForm-submit" type="submit">Edit</button>
        {formMessages.length ? <Alert type="danger" messages={formMessages} /> : null }
      </form>
    </div>
  );
};

export default EditUserForm;