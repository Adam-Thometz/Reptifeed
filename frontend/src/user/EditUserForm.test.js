import React from "react";
import { render } from '@testing-library/react'
import EditUserForm from "./EditUserForm";
import { MemoryRouter } from "react-router-dom";
import { UserProvider } from '../utils/testUtils'

test('it renders without crashing', () => {
  render(<MemoryRouter>
    <UserProvider>
      <EditUserForm />
    </UserProvider>
  </MemoryRouter>);
});

test('it matches snapshot', () => {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <EditUserForm />
      </UserProvider>
    </MemoryRouter>);
  expect(asFragment()).toMatchSnapshot();
});