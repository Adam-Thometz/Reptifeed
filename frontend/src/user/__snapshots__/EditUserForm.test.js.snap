// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="EditUserForm"
  >
    <h1>
      Edit your information
    </h1>
    <form>
      <div
        class="EditUserForm-form-group"
      >
        <label
          class="EditUserForm-label"
          for="username"
        >
          Username:
        </label>
        <input
          class="EditUserForm-input"
          id="username"
          name="username"
          type="text"
          value="rickSanchez"
        />
      </div>
      <div
        class="EditUserForm-form-group"
      >
        <label
          class="EditUserForm-label"
          for="password"
        >
          Password:
        </label>
        <input
          class="EditUserForm-input"
          id="password"
          name="password"
          type="password"
          value=""
        />
      </div>
      <div
        class="EditUserForm-form-group"
      >
        <label
          class="EditUserForm-label"
          for="email"
        >
          Email:
        </label>
        <input
          class="EditUserForm-input"
          id="email"
          name="email"
          type="email"
          value="rickman@qmail.com"
        />
      </div>
      <button
        class="EditUserForm-submit"
        type="submit"
      >
        Edit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
